### PR TITLE
Patch CVE-2023-3036 and update development environment

### PR DESCRIPTION
**Pull request by lukevalenta**

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -1,18 +1,34 @@
...
       context: .
       dockerfile: Dockerfile.memcache
+    volumes:
+      - ./scripts:/scripts
+    entrypoint: ["/scripts/run_memcached.sh"]
+  clippy:
+    build:
+      context: .
+      dockerfile: Dockerfile.clippy
+    command: ["--all-features", "--", "-D", "warnings"]
```
Comment (cjpatton) [inline docker-compose.yaml:34]: Are there unit tests in this repo to run as well?

Did you consider running cargo test && cargo clippy via GitHub's native CI stuff? @armfazh helped me set up https://github.com/cloudflare/daphne/blob/main/.github/workflows/daphneci.yml

File: src/metrics.rs [hunk 2]
```diff
@@ -39,7 +39,7 @@ fn wait_for_req_or_eof(dest: &net::TcpStream, logger: slog::Logger) -> Result<()
                 logger,
                 "failure to read request {:?}, unable to serve metrics", e
             );
-            dest.shutdown(net::Shutdown::Both);
+            let _ = dest.shutdown(net::Shutdown::Both);
```
Comment (cjpatton) [inline src/metrics.rs:42]: What is the type here? If this is a `Result` then we should consider handling it explicitly. (Here and below.)

File: src/ntp/client.rs [hunk 3]
```diff
@@ -40,9 +40,7 @@ pub enum NtpClientError {
 
 impl std::error::Error for NtpClientError {
     fn description(&self) -> &str {
-        match self {
-            _ => "Connection to server failed because address could not be resolved",
-        }
+        "Connection to server failed because address could not be resolved"
```
Comment (cjpatton) [inline src/ntp/client.rs:43]: Was this clippy? This original code looks a little strange, but I wonder if the intention was that the string returned would depend on the variant of `self`. We could do this instead:
```rust
match self {
    Self::NoIpV4AddrFound | Self::NoIpv6AddrFound | Self::InvalidUid => "XXX",
}
```
That way the compiler will force us to deal with variants explicitly.

Also, we might consider making the description actually reflect the variant :)

File: src/ntp/server/config.rs [hunk 4]
```diff
@@ -162,15 +162,11 @@ impl NtpServerConfig {
...
-        let upstream_sock_addr = if upstream_addr.is_some() && upstream_port.is_some() {
-            let sock_addr = SocketAddr::from((
-                // No problem to unwrap here because `upstream_addr` is Some(_).
-                IpAddr::from_str(&upstream_addr.unwrap()).wrap_err()?,
-
-                // No problem to unwrap here because `upstream_port` is Some(_).
-                upstream_port.unwrap(),
-            ));
-            Some(sock_addr)
+        let upstream_sock_addr = if let (Some(upstream_addr), Some(upstream_port)) = (upstream_addr, upstream_port) {
```
Comment (cjpatton) [inline src/ntp/server/config.rs:165]: Now *that's* how you write Rust!

File: src/ntp/client.rs (hunk 3 above)
Comment (cjpatton) [reply src/ntp/client.rs:43]: Similarly below.

**Review by cjpatton — APPROVED** [review]

File: docker-compose.yaml (hunk 1 above)
Comment (lukevalenta) [reply docker-compose.yaml:34]: Nice! I'll give that a try. There are some unit tests, but coverage is far from complete.

**Pull request merged**